### PR TITLE
support project.templates & openshift_url override

### LINKED ISSUE #76: Create Test and Documentation For Private app.template feature
See https://github.com/rht-labs/ansible-stacks/pull/71.

We need to:

- move the test data into it's own test
- add info to the README to indicate that this is a private feature not supported in https://github.com/rht-labs/api-design/blob/master/swagger.yaml

### LINKED ISSUE #88: The merge of PR #84 broke ansible-stacks
@etsauer the latest [PR](https://github.com/rht-labs/ansible-stacks/pull/84) broke normal operation of ansible-stacks - see below...

```
TASK [create-openshift-resources : include] ************************************
included: /home/obedin/ansible-stacks/roles/create-openshift-resources/tasks/create_environment_variables_string.yml for localhost

TASK [create-openshift-resources : Add variable to environment_variables_string] ***
fatal: [localhost]: FAILED! => {
    "failed": true
}

MSG:

The conditional check 'app.from_template is not defined and app.from_template == ''' failed. The error was: error while evaluating conditional (app.from_template is not defined and app.from_template == ''): 'dict object' has no attribute 'from_template'

The error appears to have been in '/home/obedin/ansible-stacks/roles/create-openshift-resources/tasks/create_environment_variables_string.yml': line 3, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


- name: "Add variable to environment_variables_string"
  ^ here


	to retry, use: --limit @/home/obedin/ansible-stacks/load_infra_local.retry

PLAY RECAP *********************************************************************
localhost                  : ok=39   changed=5    unreachable=0    failed=1   
```

### LINKED ISSUE #93: Model Templates As Part of Project
This feature will model templates as part of a project, not as an application. The current model is built around that applications are driven by `oc new-app`, and thus merging these two concepts is leading to conflicts as seen with #89 #88 #80. Modelling templates as part of a project allows use to clearly separate concerns in our model, as well as our implementation.

### LINKED ISSUE #94: Support Override of openshift_cluster.openshift_host_env
There is no clean or documented way to override the cluster url defined in the API doc. There are many times when you want to separate the API doc from the specific cluster it's being run in, so we need to provide a way to handle this short term.

Long term, we need to resolve https://github.com/rht-labs/api-design/issues/1

### PR DESCRIPTION
**Pull request by sherl0cks**
#### What does this PR do?
Brief explanation of the code or documentation change you've made

- Deletes all previous work done for templates and replaces it with the model introduced in https://github.com/rht-labs/api-design/pull/75. 
- Provides a documented when to set the `openshift_url` outside of the API document, or override the value in the API document
  - clean up tests to use this approach
- There is WIP committed in this PR related to #95, but it has been toggled off.
- Clean up documentation 
- prune tests for usability and to prepare for CI

#### Which tests illustrate how this code works?
See diff

#### Is there a relevant Issue open for this?
Resolves #94 
Resolves #93
Resolves #88 
Resolves #76 

#### Are there any other relevant resources that should be reviewed as well?
https://github.com/rht-labs/api-design/pull/75

#### Who would you like to review this?
/cc @oybed

### CODE REVIEW

File: roles/configure-openshift-templates/files/jenkins-ephemeral-template.json [hunk 1]
```diff
@@ -0,0 +1,275 @@
+{
```
Comment (oybed) [inline roles/configure-openshift-templates/files/jenkins-ephemeral-template.json]: I thought we were going to keep these files in the "examples" repo?

File: roles/configure-openshift-templates/files/jenkins-s2i-template.json [hunk 2]
```diff
@@ -0,0 +1,123 @@
+{
```
Comment (oybed) [inline roles/configure-openshift-templates/files/jenkins-s2i-template.json]: I thought we were going to keep these files in the "examples" repo?

File: roles/create-openshift-resources/tasks/create_templates.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  set_fact:
+    template: "{{ template_item }}"
+
+- name: Fail for Missing name
+  fail: msg="This role requires template.name to be set and non empty"
+  when: template.name is not defined or template.name == ''
+
+- name: "Determine if Objects from {{ template.name }} Exists"
+  command: >
+     {{ openshift.common.client_binary }} get ds,dc,bc,is,svc,pvc,sa,rolebindings -l pbiTemplateName={{ template.name }} -n {{ project.name }} -o json
```
Comment (oybed) [inline roles/create-openshift-resources/tasks/create_templates.yml]: in general for automation implementations, I recommend spelling out the abbreviations, such as "dc", "bc", etc. I've seen mistakes made in the past due to not doing do...

**Review by oybed — CHANGES_REQUESTED** [review]
regression testing with our delivery env .json, it fails with the following error:

```
TASK [create-openshift-resources : Set Fact for Fabric Java S2I] *******************************************************************************************************************
fatal: [localhost]: FAILED! => {"failed": true, "msg": "The conditional check 'app.labels.provider is defined and app.labels.provider == 'fabric8'' failed. The error was: error while evaluating conditional (app.labels.provider is defined and app.labels.provider == 'fabric8'): 'dict object' has no attribute 'labels'\n\nThe error appears to have been in '/home/obedin/coding/testing/ansible-stacks.templates/roles/create-openshift-resources/tasks/create_app.yml': line 100, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: \"Set Fact for Fabric Java S2I\"\n  ^ here\n"}
```

File: roles/configure-openshift-templates/files/jenkins-ephemeral-template.json (hunk 1 above)
Comment (sherl0cks) [reply roles/configure-openshift-templates/files/jenkins-ephemeral-template.json]: This was oversight, will delete.

File: roles/configure-openshift-templates/files/jenkins-s2i-template.json (hunk 2 above)
Comment (sherl0cks) [reply roles/configure-openshift-templates/files/jenkins-s2i-template.json]: This was oversight, will delete.

File: roles/create-openshift-resources/tasks/create_templates.yml (hunk 3 above)
Comment (sherl0cks) [reply roles/create-openshift-resources/tasks/create_templates.yml]: interesting, I'm happy to update accordingly.

**Comment by sherl0cks** [comment]
@oybed interesting on that regression, as there are no changes related to it, and there is also test coverage that should handle this. I'll follow up offline

**Comment by oybed** [comment]
LGTM - merging

**Review by oybed — APPROVED** [review]

**Pull request merged**